docs/design_proposals/private_pip_registries.md

```diff
@@ -0,0 +1,106 @@
+# Proposal Design: Support for Private `pip` registries
+
+## Objective
+
+The aim is to support private `pip` package registries in pixi.
+
+## Motivation
+
+To deliver and deploy internal software, organizations often host packages in private registries requiring authentication. Examples of such solutions are:
+
+* Google Cloud Artifact Registry
+* AWS CodeArtifact
+* Artifactory
+* packagecloud
+* ...etc.
+
+To use such repositories with pixi, authentication should be supported.
+
+## Design considerations
+
+1. **Compatibility**. Different package servers use different ways of deriving credentials. Some use static usernames and passwords, some use temporary tokens that need to be regenerated once in a while. Ideally, `pixi` itself should not be concerned about those details.
+2. **Ease of use**. In environments where authentication credentials can be derived automatically (e.g. GCP ADC), authentication should work "natively".
+3. **Configurable**. In `pixi.toml` file, developers should be able to choose what authentication mechanism to choose for a given channel.
+4. **Separation of concerns**. Authentication implementation for different platforms can be separate from `pixi` itself.
+
+## Proposed solution
+
+### Support for authentication helpers in `pixi`
+
+Authentication helper is a program that receives an URL to the package registry as an input and outputs a JSON object containing the credentials. Authentication helpers can utilize different methods for obtaining credentials, but it's expected that they are specific to the package host implementation.
+
+Authentication helper is a runnable binary available in `PATH`. The name of the binary must start with `pixi-auth-` and followed by its unique ID.
+
+A number of authentication helpers is shipped together with `pixi`. However, the interface between `pixi` and authentication helpers is documented, and it's expected that users may add a helper specific to their usecase.
+
+Authentication helpers are loosely inspired by `GIT_ASKPASS`, and authentication helpers in Docker and Flyte.
+
+### `pixi` interface for authentication helpers
+
+Every authentication helper program should, in case of success, return a JSON object with the following structure:
+
+```
+{
+    "username": "<username>",
+    "password": "<password>"
+}
+```
+
+Upon success, authentication helpers exit with code `0`.
+
+Upon failure, authentication helpers exit with code other than `0`. Authentication helpers must print a human-readable error message to stderr. `pixi` will relay the message back to the user.
+
+Authentication helpers are never launched in interactive mode. Authentication helpers never receive user input over `stdin`, and can't print any additional information upon success.
```

Comment:
Why not? For an oauth2 flow one might need to open a url, login and paste a token into the terminal.
This works for example in the AWS CLI https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-sourcing-external.html

Reply:
Implementating non-interactive first makes sense. I was just scared of the `never` in the sentence. 
Regarding the interface to interactive auth helpers I would argue to postpone the discussion how it looks like to a later stage. Also worth looking at how other tools do it.